### PR TITLE
Accountsdb plugin support of streaming transaction data 

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
Support streaming transaction data through the plugin interface
#### Summary of Changes
1. Expand the plugin interface to support transaction data
2. Implemented PostgreSQL types for transaction data
3. Implemented data transformation objects from SDK transaction types to the DB transaction objects
4. Streamlined the interfaces between accounts, transaction and slot notifications. Slot notification moved to its own module.
5. Unit tests
Fixes #

### CODE REVIEW

File: rpc/src/transaction_status_service.rs [hunk 1]
```diff
@@ -145,31 +152,48 @@ impl TransactionStatusService {
                                 .collect(),
                         );
 
-                        if let Some(memos) = extract_and_fmt_memos(transaction.message()) {
-                            blockstore
-                                .write_transaction_memos(transaction.signature(), memos)
-                                .expect("Expect database write to succeed: TransactionMemos");
+                        let transaction_status_meta = TransactionStatusMeta {
```
Comment (lijunwangs) [inline rpc/src/transaction_status_service.rs:155]: Used for both saving to block store and plugin interface

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -1259,7 +1259,7 @@ impl Bank {
             accounts_db_caching_enabled,
             shrink_ratio,
             accounts_db_config,
-            accounts_update_notifier,
+            accounts_update_notifier.clone(),
```
Comment (lijunwangs) [inline runtime/src/bank.rs:1262]: Review changes in this module and revert if unnecessary

File: runtime/src/serde_snapshot.rs [hunk 3]
```diff
@@ -354,7 +354,7 @@ where
         shrink_ratio,
         verify_index,
         accounts_db_config,
-        accounts_update_notifier,
+        accounts_update_notifier.clone(),
```
Comment (lijunwangs) [inline runtime/src/serde_snapshot.rs]: revert unnecessary clone

File: runtime/src/bank.rs (hunk 2 above)
Comment (lijunwangs) [reply runtime/src/bank.rs]: Remved

File: accountsdb-plugin-interface/src/accountsdb_plugin_interface.rs [hunk 4]
```diff
@@ -79,21 +96,53 @@ pub trait AccountsDbPlugin: Any + Send + Sync + std::fmt::Debug {
...
-    ) -> Result<()>;
+    ) -> Result<()> {
+        Ok(())
+    }
+
+    /// Called when a transaction is updated at a slot.
+    #[allow(unused_variables)]
+    fn notify_transaction(
+        &mut self,
+        transaction: ReplicaTranscaionInfoVersions,
```
Comment (carllin) [inline accountsdb-plugin-interface/src/accountsdb_plugin_interface.rs:129]: nit: `ReplicaTranscaionInfoVersions -> ReplicaTransactionInfoVersions`

File: accountsdb-plugin-postgres/src/accounts_selector.rs [hunk 5]
```diff
@@ -48,6 +48,11 @@ impl AccountsSelector {
     pub fn is_account_selected(&self, account: &[u8], owner: &[u8]) -> bool {
         self.select_all_accounts || self.accounts.contains(account) || self.owners.contains(owner)
     }
+
+    /// Check if any account is of interested at all
+    pub fn is_interested_in_any_account(&self) -> bool {
```
Comment (carllin) [inline accountsdb-plugin-postgres/src/accounts_selector.rs:53]: nit: `is_interested_in_any_account` -> `is_enabled()`

File: accountsdb-plugin-postgres/src/transaction_selector.rs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        for account in accounts {
+            if self.mentions.contains(account.as_ref()) {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Check if any transaction is of interested at all
+    pub fn is_interested_in_any_transaction(&self) -> bool {
```
Comment (carllin) [inline accountsdb-plugin-postgres/src/transaction_selector.rs:74]: nit: is_interested_in_any_transaction -> is_enabled()

File: accountsdb-plugin-postgres/src/accountsdb_plugin_postgres.rs [hunk 7]
```diff
@@ -276,6 +296,61 @@ impl AccountsDbPlugin for AccountsDbPluginPostgres {
...
+
+    /// Check if the plugin is interested in account data
+    fn is_account_data_interested(&self) -> bool {
+        self.accounts_selector
+            .as_ref()
+            .map_or_else(|| false, |selector| selector.is_interested_in_any_account())
+    }
+
+    /// Check if the plugin is interested in transaction data
+    fn is_transaction_data_interested(&self) -> bool {
```
Comment (carllin) [inline accountsdb-plugin-postgres/src/accountsdb_plugin_postgres.rs:348]: nit: `is_transaction_data_interested -> is_transaction_storage_enabled`

File: accountsdb-plugin-postgres/src/accountsdb_plugin_postgres.rs [hunk 8]
```diff
@@ -276,6 +296,61 @@ impl AccountsDbPlugin for AccountsDbPluginPostgres {
...
+                    }
+                }
+            },
+        }
+
+        Ok(())
+    }
+
+    /// Check if the plugin is interested in account data
+    fn is_account_data_interested(&self) -> bool {
```
Comment (carllin) [inline accountsdb-plugin-postgres/src/accountsdb_plugin_postgres.rs:341]: nit: `is_account_data_interested -> is_account_storage_enabled`

**Comment by carllin** [comment]
This PR looks like it can be broken up into a few separate/more modular PR's for easier review, especially given the amount of refactoring + plumbing + new code changes:

1) The `SlotStatusNotifierInterface` refactoring changes
2) The `TransactionSelector` changes
3) The interface definitions in `transaction_notifier_interface.rs`, `transaction_notifier.rs`, `accountsdb_plugin_postgres.rs` with a stub for the implementation of `fn log_transaction_info()` in `accountsdb_plugin_postgres.rs`
4) The set of big changes and actual definition of `log_transaction_info()` and its associated utilities in `postgres_client_transaction.rs`

**Comment by jstarry** [comment]
> This PR looks like it can be broken up into a few separate/more modular PR's for easier review, especially given the amount of refactoring + plumbing + new code changes:

Agreed, it's hard to find the time for such a large review

**Comment by lijunwangs** [comment]
> This PR looks like it can be broken up into a few separate/more modular PR's for easier review, especially given the amount of refactoring + plumbing + new code changes:
> 
> 1. The `SlotStatusNotifierInterface` refactoring changes
> 2. The `TransactionSelector` changes
> 3. The interface definitions in `transaction_notifier_interface.rs`, `transaction_notifier.rs`, `accountsdb_plugin_postgres.rs` with a stub for the implementation of `fn log_transaction_info()` in `accountsdb_plugin_postgres.rs`
> 4. The set of big changes and actual definition of `log_transaction_info()` and its associated utilities in `postgres_client_transaction.rs`

okay -- let me look into splitting it into smaller PRs.

**Comment by lijunwangs** [comment]
Closed for now as I will split it into smaller PRs

**Pull request closed without merging**